File: it-agents/skills/code-architect/references/code-examples/java/Iterator.java
// Iterator — traverse a collection without exposing its representation.

interface SequenceIterator<T> {
    boolean hasNext();
    T next();
}

interface Sequence<T> {
    SequenceIterator<T> iterator();
}

// Custom collection backed by a ring buffer, hiding its internals.
class RingBuffer<T> implements Sequence<T> {
    private final Object[] data;
    private int count = 0;
    RingBuffer(int capacity) { data = new Object[capacity]; }
    void add(T item) { data[count++ % data.length] = item; }

    public SequenceIterator<T> iterator() {
        return new SequenceIterator<T>() {
            private int pos = 0;
            public boolean hasNext() { return pos < count; }
            @SuppressWarnings("unchecked")
            public T next() { return (T) data[pos++ % data.length]; }
        };
    }
}

public class Iterator {
    public static void main(String[] args) {
        RingBuffer<String> buffer = new RingBuffer<>(8);
        buffer.add("a");
        buffer.add("b");
        buffer.add("c");

        SequenceIterator<String> it = buffer.iterator();
        while (it.hasNext()) {
            System.out.println("item: " + it.next());
        }
    }
}

File: it-agents/skills/code-architect/references/code-examples/typescript/Iterator.ts
// Iterator — traverse a collection without exposing its representation.

interface SimpleIterator<T> {
  hasNext(): boolean;
  next(): T;
}

// Aggregate backed by a ring buffer; iteration order is hidden.
class RingBuffer<T> {
  private items: T[] = [];

  push(item: T): void {
    this.items.push(item);
  }

  // Iterates in reverse insertion order without revealing internals.
  createIterator(): SimpleIterator<T> {
    let cursor = this.items.length - 1;
    const snapshot = [...this.items];
    return {
      hasNext: () => cursor >= 0,
      next: () => snapshot[cursor--],
    };
  }
}

function demo(): void {
  const buffer = new RingBuffer<string>();
  buffer.push("a");
  buffer.push("b");
  buffer.push("c");

  const it = buffer.createIterator();
  const visited: string[] = [];
  while (it.hasNext()) {
    visited.push(it.next());
  }
  console.log(`visited: ${visited.join(" -> ")}`); // c -> b -> a
}

demo();
